eth/tracers/internal/tracers/call_tracer.js

```diff
@@ -0,0 +1,246 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// callTracer is a full blown transaction tracer that extracts and reports all
+// the internal calls made by a transaction, along with any useful information.
+{
+	// callstack is the current recursive call stack of the EVM execution.
+	callstack: [{}],
+
+	// descended tracks whether we've just descended from an outer transaction into
+	// an inner call.
+	descended: false,
+
+	// step is invoked for every opcode that the VM executes.
+	step: function(log, db) {
+		// Capture any errors immediately
+		var error = log.getError();
+		if (error !== undefined) {
+			this.fault(log, db);
+			return;
+		}
+		// We only care about system opcodes, faster if we pre-check once
+		var syscall = (log.op.toNumber() & 0xf0) == 0xf0;
+		if (syscall) {
+			var op = log.op.toString();
```

Comment:
Since 'op' will be undefined if it's not an 0xf* opcode, why not put the entire following section in an if, or break out early?

Reply:
`op` is accessed  various places and there are code segments in between that do not depend on `op` and still need to run even if `op` is undefined.

Reply:
`op` will be undefined if `syscall` is false, so you could at least simplify the following checks to not check both.

Reply:
My hunch was that evaluating whether `syscall` is false or not is a boolean operation and should be faster than mucking around with `op`. Running both codes for the start of Ropsten, there's indeed a slight performance increase for the current code versus the one that doesn't check for `syscall`:

Check both:

```
start=0 end=36863 current=22126 transactions=5027 elapsed=1m21.076843668s
```

vs. check only op:

```
start=0 end=36863 current=21778 transactions=4844 elapsed=1m20.62278927s
```

Reply:
Not much, about 4% difference.